platform/rpc/nats/bus.go

```diff
@@ -0,0 +1,234 @@
+	var response RegradeResponse
+	if err := r.conn.Request(r.instance+methodRegrade, specs, &response, timeout); err != nil {
+		return err
+	}
+	if len(response.Result) > 0 {
+		errs := platform.RegradeError{}
+		for s, e := range response.Result {
+			errs[s] = errors.New(e)
+		}
+		return errs
+	}
+	return maybeError(response.Error)
+}
+
+func (r *requester) Ping() error {
+	var response ping
+	return r.conn.Request(r.instance+methodPing, ping{}, &response, timeout)
+}
+
+// Connect returns a platform.Platform implementation that can be used
+// to talk to a particular instance.
+func (n *NATS) Connect(instID flux.InstanceID) (platform.Platform, error) {
+	return &requester{
+		conn:     n.snd,
+		instance: string(instID),
+	}, nil
+}
+
+// Subscribe registers a remote platform.Platform implementation as
+// representing a particular instance ID, blocking indefinitely.
+func (n *NATS) Subscribe(instID flux.InstanceID, remote platform.Platform, done chan<- error) {
+	encoder := nats.EncoderForType(encoderType)
+
+	requests := make(chan *nats.Msg)
+	sub, err := n.rcv.ChanSubscribe(string(instID)+".Platform.>", requests)
+	if err != nil {
+		done <- err
+		return
+	}
+
+	go func() {
+		var err error
+		for request := range requests {
+			switch {
+			case strings.HasSuffix(request.Subject, methodPing):
+				var p ping
+				if err = encoder.Decode(request.Subject, request.Data, &p); err != nil {
+					break
+				}
+				n.snd.Publish(request.Reply, ping{})
+			case strings.HasSuffix(request.Subject, methodAllServices):
+				var (
+					req rpc.AllServicesRequest
+					res []platform.Service
+				)
+				err = encoder.Decode(request.Subject, request.Data, &req)
+				if err == nil {
+					res, err = remote.AllServices(req.MaybeNamespace, req.Ignored)
+				}
+				n.snd.Publish(request.Reply, AllServicesResponse{res, maybeString(err)})
+			case strings.HasSuffix(request.Subject, methodSomeServices):
+				var (
+					req []flux.ServiceID
+					res []platform.Service
+				)
+				err = encoder.Decode(request.Subject, request.Data, &req)
+				if err == nil {
+					res, err = remote.SomeServices(req)
+				}
+				n.snd.Publish(request.Reply, SomeServicesResponse{res, maybeString(err)})
+			case strings.HasSuffix(request.Subject, methodRegrade):
+				var (
+					req []platform.RegradeSpec
+				)
+				err = encoder.Decode(request.Subject, request.Data, &req)
+				if err == nil {
+					err = remote.Regrade(req)
+				}
+				response := RegradeResponse{}
+				switch regradeErr := err.(type) {
+				case platform.RegradeError:
+					result := rpc.RegradeResult{}
+					for s, e := range regradeErr {
+						result[s] = e.Error()
+					}
+					response.Result = result
+				default:
+					response.Error = maybeString(err)
+				}
+				n.snd.Publish(request.Reply, response)
+			default:
+				err = errors.New("unknown message: " + request.Subject)
+			}
+			if err != nil {
+				sub.Unsubscribe()
+				close(requests)
+				done <- err
+				return
+			}
+		}
+	}()
```

Comment:
This giant switch *might* be easier to clean up if you separate out the "decode the thing", from the "do the thing" phases. Not sure.